cosmwasm/contracts/ntt-global-accountant/src/contract.rs

```diff
@@ -0,0 +1,962 @@
+        match handle_observation(deps.branch(), o, guardian_set_index, quorum, signature) {
+            Ok((status, event)) => {
+                responses.push(SubmitObservationResponse { key, status });
+                if let Some(evt) = event {
+                    events.push(evt);
+                }
+            }
+            Err(e) => {
+                let err = ObservationError {
+                    key,
+                    error: format!("{e:#}"),
+                };
+                let evt = cw_transcode::to_event(&err)
+                    .context("failed to transcode observation error")?;
+                events.push(evt);
+                responses.push(SubmitObservationResponse {
+                    key: err.key,
+                    status: ObservationStatus::Error(err.error),
+                });
+            }
+        }
+    }
+
+    let data = to_binary(&responses).context("failed to serialize transfer details")?;
+
+    Ok(Response::new()
+        .add_attribute("action", "submit_observations")
+        .add_attribute("owner", info.sender)
+        .set_data(data)
+        .add_events(events))
+}
+
+fn handle_observation(
+    mut deps: DepsMut<WormholeQuery>,
+    o: Observation,
+    guardian_set_index: u32,
+    quorum: u32,
+    sig: Signature,
+) -> anyhow::Result<(ObservationStatus, Option<Event>)> {
+    let relayer_emitter = RELAYER_CHAIN_REGISTRATIONS.may_load(deps.storage, o.emitter_chain)?;
+
+    let digest = o.digest().context(ContractError::ObservationDigest)?;
+    let event = cw_transcode::to_event(&o)
+        .map(Some)
+        .context("failed to transcode `Observation` to `Event`")?;
+
+    let (sender, payload) =
+        if relayer_emitter.is_some_and(|relayer_address| relayer_address == o.emitter_address) {
+            // if the emitter is a known standard relayer, parse the sender and payload from the delivery instruction
+            let delivery_instruction = DeliveryInstruction::deserialize(&o.payload.0)?;
+            (
+                delivery_instruction.sender_address.into(),
+                delivery_instruction.payload,
+            )
+        } else {
+            // otherwise, the sender and payload is the same as the VAA
+            (o.emitter_address.into(), o.payload.0)
+        };
+
+    let hub_key = TRANSCEIVER_TO_HUB.key((o.emitter_chain, sender));
+
+    let hub = hub_key
+        .may_load(deps.storage)
+        .context("failed to load hub")?
+        .ok_or(ContractError::MissingHubRegistration)?;
+
+    let message: TransceiverMessage<WormholeTransceiver, NativeTokenTransfer> =
+        TypePrefixedPayload::read_payload(&mut payload.as_slice())
+            .context("failed to parse observation payload")?;
+
+    let destination_chain = message.ntt_manager_payload.payload.to_chain.id;
+    let source_peer_key = TRANSCEIVER_PEER.key((o.emitter_chain, sender, destination_chain));
+    let source_peer = source_peer_key
+        .may_load(deps.storage)
```

Comment:
`may_load` is used a lot when there is no case to handle the option in a special way (like with attempting to fetch the relayer emitter above). These all can be changed to `load` and mapping to a specific error.